web-server/plugins/slycat-parameter-image/js/Components/Histogram.tsx

```diff
@@ -0,0 +1,283 @@
+import React, { useEffect, useRef } from "react";
+import * as d3 from "d3v7";
+import { SlycatScaleType, ValueIndexType } from "../selectors";
+import slycat_color_maps from "js/slycat-color-maps";
+
+type HistogramProps = {
+  x_scale: SlycatScaleType;
+  y_scale: SlycatScaleType;
+  y_scale_range: number[];
+  bins: d3.Bin<ValueIndexType, number>[];
+  selected_bin_indexes: number[];
+  partially_selected_bin_indexes: number[];
+  flash_bin_indexes: number[];
+  x_ticks: number;
+  y_ticks: number;
+  histogram_bar_color: string;
+  histogram_bar_stroke_width: number;
+  histogram_bar_selected_stroke_width: number;
+  font_size: number;
+  font_family: string;
+  x_label_y: number;
+  x_label_x: number;
+  x_name: string;
+  y_label_y: number;
+  colormap: string;
+  y_label_horizontal_offset: number;
+  handleBinClick: (
+    event: React.MouseEvent<SVGRectElement, MouseEvent>,
+    bin: {
+      range: (number | undefined)[];
+      count: number;
+      index: number;
+      bins_length: number;
+      data: d3.Bin<ValueIndexType, number>;
+    },
+  ) => void;
+  handleBackgroundClick: (event: React.MouseEvent) => void;
+  tickFormatter: (value: number) => string;
+};
+
+const Histogram: React.FC<HistogramProps> = (props) => {
+  const histogramRef = useRef<SVGSVGElement>(null);
+  const MIN_BAR_DIMENSIONS = 1;
+
+  const {
+    colormap,
+    x_scale,
+    y_scale_range,
+    x_ticks,
+    y_ticks,
+    histogram_bar_stroke_width,
+    histogram_bar_selected_stroke_width,
+    font_size,
+    font_family,
+    x_label_y,
+    x_label_x,
+    x_name,
+    y_label_y,
+    bins,
+    selected_bin_indexes,
+    partially_selected_bin_indexes,
+    flash_bin_indexes,
+    y_scale,
+    histogram_bar_color,
+    y_label_horizontal_offset,
+    handleBinClick,
+    handleBackgroundClick,
+    tickFormatter,
+  } = props;
+
+  // If x_scale is a band scale, set the padding to 0.1 to create
+  // spacing between bars.
+  x_scale.bandwidth && x_scale.padding(0.1);
+
+  // Flash the selected bins
+  const doFlashBins = () => {
+    // Select the histogram
+    const histogram = d3.select(histogramRef.current);
+    // Select all the bars
+    const bars = histogram.selectAll("rect");
+    // Set the duration of the transition
+    const duration = 500;
+    // Set the delay of the transition
+    const delay = 0;
+    // Set the color of the flash
+    const flash_color = "red";
+    // Set the opacity of the flash
+    const flash_opacity = 0.5;
+    // Flash the bins
+    flash_bin_indexes.forEach((binIndex) => {
+      bars
+        .filter((d, i) => i === binIndex)
+        .transition()
+        .duration(duration)
+        .delay(delay)
+        .style("fill", flash_color)
+        .style("opacity", flash_opacity)
+        .transition()
+        .duration(duration)
+        .delay(delay)
+        .style("fill", "")
+        .style("opacity", "");
+    });
+  };
+
+  // This effect creates the histogram using d3.
+  useEffect(() => {
```

Comment:
why is this in a useEffect?

Reply:
We need to run d3 code each time the React component is mounted and updated. For a functional component, `componentDidMount` and `componentDidUpdate` are not available. So the recommendation is to use `useEffect` to run that code.